### PR TITLE
feat(mutators) Avoid creating some unnecessary mutators for (a && b) …

### LINKED ISSUE #3339: Avoid using "true" condition replacement in "||" statements, and avoid using "false" replacement in "&&" statements
**Is your feature request related to a problem? Please describe.**

I have some code like this : 

```
if (part.type === "content" || part.type === "tag") {
    return part;
}
```

And there are 3 applied mutants : 

```
if (true) {
    return part;
}

if (true || part.type === "tag") {
    return part;
}
if (part.type === "content" || true) {
    return part;
}
```

I think it would make sense to just run one of the mutants in this case.

### PR DESCRIPTION
**Pull request by edi9999**
Fix #3339

Will be adding a few comments inline

Before:

```diff
- if (a && b) {}
+ if (false) {}
+ if (false && b) {}
// And others

- if (a || b) {}
+ if (true) {}
+ if (true && b) {}
// And others
```

After:

```diff
- if (a && b) {}
+ if (false) {}
// And others

- if (a || b) {}
+ if (true) {}
// And others
```

### CODE REVIEW

File: packages/instrumenter/test/unit/mutators/conditional-expression-mutator.spec.ts [hunk 1]
```diff
@@ -70,6 +70,61 @@ describe(sut.name, () => {
...
+      sut,
+      'if ((c1 || c2) && (c3 || c4)) { a++ }',
+      'if (true) { a++ }',
+      'if (false) { a++ }',
+      'if ((true) && (c3 || c4)) { a++ }',
+      'if ((c1 || c2) && (true)) { a++ }'
+    );
+  });
+
+  it('should mutate (c1 || c2 || c3)', () => {
```
Comment (nicojs) [inline packages/instrumenter/test/unit/mutators/conditional-expression-mutator.spec.ts:117]: This test ist failing in CI

File: packages/instrumenter/test/helpers/expect-mutation.ts [hunk 2]
```diff
@@ -52,6 +52,10 @@ export function expectJSMutation(sut: NodeMutator, originalCode: string, ...expe
       }
     },
   });
-  expect(mutants).lengthOf(expectedReplacements.length);
-  expectedReplacements.forEach((expected) => expect(mutants, `was: ${mutants.join(',')}`).to.include(expected));
+  /* eslint-disable @typescript-eslint/require-array-sort-compare */
+  /* because we know mutants and expectedReplacements are strings */
+  mutants.sort();
```
Comment (nicojs) [inline packages/instrumenter/test/helpers/expect-mutation.ts:57]: Good idea!

File: packages/instrumenter/src/mutators/conditional-expression-mutator.ts [hunk 3]
```diff
@@ -14,6 +14,22 @@ export const conditionalExpressionMutator: NodeMutator = {
       yield types.booleanLiteral(true);
       yield types.booleanLiteral(false);
     } else if (isBooleanExpression(path)) {
+      if (path.parent && path.parent.type === 'LogicalExpression') {
```
Comment (nicojs) [inline packages/instrumenter/src/mutators/conditional-expression-mutator.ts:17]: Can we use optional chaining here? `path.parent?.type === 'LogicalExpression'`

**Review by nicojs — CHANGES_REQUESTED** [review]
Looks incredibly simple if you put it like this 🙂

Got some small remarks. Also CI is failing

**Comment by nicojs** [comment]
> Maybe this test should simply be removed from this pull request, and a new issue created for this issue.

Yes, that's probably best. Probably a good idea to discuss it in a separate issue as well, as I think the test cases are already covered largely by the `logicalOperatorMutator`

**Comment by edi9999** [comment]
@nicojs  is there a way to update the e2e tests quickly ?

I'm running : 

```
npm run e2e
```

Since it stops after the first fail, I have to manually relaunch the tests and copy paste the output.

**Comment by nicojs** [comment]
You can run them manually by running `npm test` in an e2e/test/xxx directory. But every time you change something in Stryker, you need to reinstall with `npm install` from the e2e root dir

**Comment by edi9999** [comment]
Tests are now passing, can you please review @nicojs ?

**Review by nicojs — APPROVED** [review]
LGTM 👍

**Pull request merged**